updater/src/updater.rs

```diff
@@ -248,189 +301,406 @@ impl Updater {
+
+							if update_block_number > current_block_number {
+								info!(target: "updater", "Update for binary {} will be triggered at block {}", binary, update_block_number);
+							} else {
+								// will lock self.state
+								drop(state);
+								self.updater_step();
+							}
+						}
+					}
+				},
 			}
-			Ok(())
-		})().unwrap_or_else(|(e, fatal)| { self.state.lock().disabled = fatal; warn!("{}", e); });
+		}
 	}
 
 	fn poll(&self) {
 		trace!(target: "updater", "Current release is {} ({:?})", self.this, self.this.hash);
 
 		// We rely on a secure state. Bail if we're unsure about it.
-		if self.client.upgrade().map_or(true, |s| !s.chain_info().security_level().is_full()) {
+		if self.client.upgrade().map_or(true, |c| !c.chain_info().security_level().is_full()) {
 			return;
 		}
 
-		let current_number = self.client.upgrade().map_or(0, |c| c.block_number(BlockId::Latest).unwrap_or(0));
-
-		let mut capability = CapState::Unknown;
-		let latest = self.collect_latest().ok();
-		if let Some(ref latest) = latest {
-			trace!(target: "updater", "Latest release in our track is v{} it is {}critical ({} binary is {})",
-				latest.track.version,
-				if latest.track.is_critical {""} else {"non-"},
-				&platform(),
-				if let Some(ref b) = latest.track.binary {
-					format!("{}", b)
-				} else {
-					"unreleased".into()
-				}
-			);
-			let mut s = self.state.lock();
-			let running_later = latest.track.version.version < self.version_info().version;
-			let running_latest = latest.track.version.hash == self.version_info().hash;
-			let already_have_latest = s.installed.as_ref().or(s.ready.as_ref()).map_or(false, |t| *t == latest.track);
-
-			if !s.disabled && self.update_policy.enable_downloading && !running_later && !running_latest && !already_have_latest {
-				if let Some(b) = latest.track.binary {
-					if s.fetching.is_none() {
-						if self.updates_path(&Self::update_file_name(&latest.track.version)).exists() {
-							info!(target: "updater", "Already fetched binary.");
-							s.fetching = Some(latest.track.clone());
-							drop(s);
-							self.fetch_done(Ok(PathBuf::new()));
-						} else {
-							if s.backoff.iter().all(|&(_, instant)| Instant::now() >= instant) {
-								info!(target: "updater", "Attempting to get parity binary {}", b);
-								s.fetching = Some(latest.track.clone());
-								drop(s);
-								let weak_self = self.weak_self.lock().clone();
-								let f = move |r: Result<PathBuf, fetch::Error>| if let Some(this) = weak_self.upgrade() { this.fetch_done(r) };
-								let a = fetch::Abort::default().with_max_size(self.update_policy.max_size);
-								self.fetcher.fetch(b, a, Box::new(f));
-							}
-						}
-					}
-				}
+		// Only check for updates every n blocks
+		let current_block_number = self.client.upgrade().map_or(0, |c| c.block_number(BlockId::Latest).unwrap_or(0));
+		if current_block_number % cmp::max(self.update_policy.frequency, 1) != 0 {
```

Comment:
How often do we `poll`? What's the guarantee that it will ever be called?

Reply:
This is currently driven by `ChainNotify` and we only `poll` after we're done syncing.

Reply:
Ok! I was worried that if it's run by a timer instead it might go into some weird stalled situations where it's aligned with multiplication of block time and the frequency would work incorrectly.